artmanflow/steps/java_sources_staging_guest.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+import os
+
+import github3
+
+try:
+    from common import GuestStepProperties, ConfigUtils, GitUtils, BaseGuest
+except ImportError:
+    from artmanflow.steps.common import \
+        GuestStepProperties, ConfigUtils, GitUtils, BaseGuest
+
+
+class JavaSourcesStagingGuest(BaseGuest):
+    def execute(self):
+        try:
+            self.before_execute()
+            staging_name = self.checkout_git_output_repo(
+                self._config['staging'])
+            client_folders = self._extract_client_folders()
+            self._copy_artifacts_to_staging(staging_name, client_folders)
+            self._fix_staging_versions(staging_name)
+            self._build_and_test(staging_name)
+            self._git_commit_and_push(staging_name)
+            pr_url = self._post_pr()
+            self._dump_output({'pr_url': pr_url})
+            self.after_execute(self._config['debug_mode'])
+        except Exception as e:
+            self.after_execute(True, e)
+            raise
+
+    def _extract_client_folders(self):
+        exp = re.compile(r'^\./([^/]+/){2}$')
+        art_name = self._config['generator_artifacts']['sources_zip']
+
+        art_list = self.check_command(['tar', '-tf', art_name],
+                                      self._guest.guest_root_path())
+        art_list = art_list.decode('UTF-8').split('\n')
+        client_folders = []
+        for art in art_list:
+            if exp.match(art):
+                client_folders.append(art)
+
+        return client_folders
+
+    def _copy_artifacts_to_staging(self, staging_name, client_folders):
```

Comment:
We have a lot of this code elsewhere; how is this different. Can the code housed elsewhere be generalized?

Reply:
One fundamental difference is that it supports batch copy and uses convention over configuration approach (does not require individual pathes to be configured in `artman_config.yaml`, but just expect the source and dest directories follow naming and folder structure convention). This simplifies things a lot (at least for java).